File: 백준/Silver/10974. 모든 순열/모든 순열.java
import java.io.*;

public class Main {
    private static int size = 0;
    private static StringBuilder sb = new StringBuilder();
    private static boolean[] isVisited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        size = Integer.parseInt(br.readLine());
        isVisited = new boolean[size + 1];
        back_tracking(1, 0);
    }

    private static void back_tracking(int num, int depth) {
        if (depth == size) {
            System.out.println(sb);
            return;
        }

        for (int i = 1; i <= size; i++) {
            if (!isVisited[i]) {
                isVisited[i] = true;
                sb.append(i).append(" ");
                back_tracking(i + 1, depth + 1);
                isVisited[i] = false;
                sb.delete(sb.length() - 2, sb.length());
            }
        }
    }
}

File: 백준/Silver/10974. 모든 순열/모든 순열.py
import sys

size = int(sys.stdin.readline())
answer = []
isVisited = [False] * (size + 1)


def back_tracking(num: int, depth: int):
    global size
    if size == depth:
        print(' '.join(map(str, answer)))
        return

    for i in range(1, size + 1):
        if not isVisited[i]:
            isVisited[i] = True
            answer.append(i)
            back_tracking(i + 1, depth + 1)
            isVisited[i] = False
            del answer[-1]


back_tracking(1, 0)
